thoth-app/src/service/account.rs

```diff
@@ -47,8 +62,110 @@ impl AccountService {
         self.update_storage(None)
     }
 
-    pub fn redirect_to_login(&self) {
-        let mut router: RouteAgentDispatcher<()> = RouteAgentDispatcher::new();
-        router.send(RouteRequest::ChangeRoute(self.login_route.clone()))
+    pub fn login(
+        &mut self,
+        login_credentials: LoginCredentials,
+        callback: Callback<Result<AccountDetails, AccountError>>,
+    ) -> FetchTask {
+        self.post_request::<LoginCredentials, AccountDetails>(
+            "/account/login".to_string(),
+            login_credentials,
+            callback,
+        )
+    }
+
+    pub fn renew_token(
+        &mut self,
+        callback: Callback<Result<AccountDetails, AccountError>>,
+    ) -> FetchTask {
+        self.bodyless_post_request::<AccountDetails>("/account/token/renew".to_string(), callback)
+    }
+
+    pub fn account_details(
+        &mut self,
+        callback: Callback<Result<AccountDetails, AccountError>>,
+    ) -> FetchTask {
+        self.get_request::<AccountDetails>("/account".to_string(), callback)
+    }
+
+    fn request_builder<B, T>(
+        &mut self,
+        method: &str,
+        url: String,
+        body: B,
+        callback: Callback<Result<T, AccountError>>,
+    ) -> FetchTask
+    where
+        for<'de> T: Deserialize<'de> + 'static + std::fmt::Debug,
+        B: Into<Text> + std::fmt::Debug,
+    {
+        let handler = move |response: Response<Text>| {
+            if let (meta, Ok(data)) = response.into_parts() {
+                if meta.status.is_success() {
+                    let data: Result<T, _> = serde_json::from_str(&data);
+                    if let Ok(data) = data {
+                        callback.emit(Ok(data))
+                    } else {
+                        callback.emit(Err(AccountError::ResponseError))
+                    }
+                } else {
+                    match meta.status.as_u16() {
+                        401 => callback.emit(Err(AccountError::AuthenticationError)),
```

Comment:
Consider making constants for these HTTP status codes

Reply:
Good point - I'll update these